libutils/threaded_queue.c

```diff
@@ -0,0 +1,504 @@
+        free(queue->cond_empty);
+        free(queue->cond_non_empty);
+        free(queue);
+        return NULL;
+    }
+
+    queue->capacity = initial_capacity;
+    queue->head = 0;
+    queue->tail = 0;
+    queue->size = 0;
+    queue->data = xmalloc(sizeof(void *) * initial_capacity);
+    queue->ItemDestroy = ItemDestroy;
+
+    return queue;
+}
+
+void ThreadedQueueDestroy(ThreadedQueue *queue)
+{
+    if (queue != NULL)
+    {
+        ThreadLock(queue->lock);
+        DestroyRange(queue, queue->head, queue->tail);
+        ThreadUnlock(queue->lock);
+
+        ThreadedQueueSoftDestroy(queue);
+    }
+}
+
+void ThreadedQueueSoftDestroy(ThreadedQueue *queue)
+{
+    if (queue != NULL)
+    {
+        if (queue->lock != NULL)
+        {
+            pthread_mutex_destroy(queue->lock);
+            free(queue->lock);
+        }
+
+        if (queue->cond_non_empty != NULL)
+        {
+            pthread_cond_destroy(queue->cond_non_empty);
+            free(queue->cond_non_empty);
+        }
+
+        if (queue->cond_empty != NULL)
+        {
+            pthread_cond_destroy(queue->cond_empty);
+            free(queue->cond_empty);
+        }
+
+        free(queue->data);
+        free(queue);
+    }
+}
+
+void *ThreadedQueuePop(ThreadedQueue *queue, int timeout)
+{
+    assert(queue != NULL);
+
+    ThreadLock(queue->lock);
+
+    if (queue->size == 0 && timeout != 0)
+    {
+        int res = 0;
+        do {
+            res = ThreadWait(queue->cond_non_empty, queue->lock, timeout);
+
+            // If ThreadWait has timed out, thread not acquired
+            if (res != 0)
+            {
+                ThreadUnlock(queue->lock);
+                return NULL;
+            }
+        } while (queue->size == 0);
+        // Reevaluate predicate to protect against spurious wakeups
+    }
+
+    void *item = NULL;
+
+    if (queue->size > 0)
+    {
+        size_t head = queue->head;
+        item = queue->data[head];
+
+        queue->data[head++] = NULL;
+
+        head %= queue->capacity;
+        queue->head = head;
+        queue->size--;
+    }
+
+    if (queue->size == 0)
+    {
+        // Signals that the queue is empty for ThreadedQueueWaitEmpty
+        pthread_cond_broadcast(queue->cond_empty);
+    }
+
+    ThreadUnlock(queue->lock);
+
+    return item;
+}
+
+size_t ThreadedQueuePopN(ThreadedQueue *queue,
+                         void ***data_array,
+                         size_t num,
+                         int timeout)
+{
+    assert(queue != NULL);
+
+    ThreadLock(queue->lock);
+
+    if (queue->size == 0 && timeout != 0)
+    {
+        int res = 0;
+        do {
+            res = ThreadWait(queue->cond_non_empty, queue->lock, timeout);
+
+            // If ThreadWait has timed out, thread not acquired
+            if (res != 0)
+            {
+                ThreadUnlock(queue->lock);
+                *data_array = NULL;
+                return 0;
+            }
+        } while (queue->size == 0);
```

Comment:
I think it's possible that predicate rechecking can result in race conditions, since another asynchronous operation might empty the queue before this thread acquires the lock, which will result in the thread just continuing to wait.

Reply:
> But the lock is acquired as part of the ThreadWait() call, right? So I don't think this point can ever be reached while not holding the lock.

I meant that: even though the thread has been signalled, the lock might be acquired by another thread before the signalled thread acquires it, which means that the predicate might be changed before this thread ever gets to check it, resulting in it just looping and waiting again. I suppose that's fine for ThreadedQueuePop, but it is probably _not_ fine for ThreadedQueueWaitEmpty, since you might have several threads that want to populate the queue when it has become empty.
I don't think there's a better way to protect against random wake-ups, so it will have to do.

Reply:
> I meant that: even though the thread has been signalled, the lock might be acquired by another thread before the signalled thread acquires it

The thread *being signalled* and the thread *acquiring the lock* happens at the same time, AFAICT.